sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasStateServer.scala

```diff
@@ -158,11 +175,70 @@ class TransformWithStateInPandasStateServer(
           Some(message.getGetValueState.getTtl.getDurationMs)
         } else None
         initializeValueState(stateName, schema, ttlDurationMs)
+      case StatefulProcessorCall.MethodCase.UTILSCALL =>
+        handleStatefulProcessorUtilRequest(message.getUtilsCall)
       case _ =>
         throw new IllegalArgumentException("Invalid method call")
     }
   }
 
+  private def handleStatefulProcessorUtilRequest(message: UtilsCallCommand): Unit = {
+    message.getMethodCase match {
+      case UtilsCallCommand.MethodCase.ISFIRSTBATCH =>
+        if (!hasInitialState) {
+          // In physical planning, hasInitialState will always be flipped
+          // if it is not first batch
+          sendResponse(1)
+        } else {
+          sendResponse(0)
+        }
+
+      case UtilsCallCommand.MethodCase.GETINITIALSTATE =>
+        if (!hasInitialState || initialStateKeyToRowMap.isEmpty) {
+          sendResponse(1)
+        } else {
+          sendResponse(0)
+
+          outputStream.flush()
+          val arrowStreamWriter = {
+            val outputSchema = initialStateSchema
+            val arrowSchema = ArrowUtils.toArrowSchema(outputSchema, timeZoneId,
+              errorOnDuplicatedFieldNames, largeVarTypes)
+            val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+              s"stdout writer for transformWithStateInPandas state socket", 0, Long.MaxValue)
+            val root = VectorSchemaRoot.create(arrowSchema, allocator)
+            new BaseStreamingArrowWriter(root, new ArrowStreamWriter(root, null, outputStream),
+              arrowTransformWithStateInPandasMaxRecordsPerBatch)
+          }
+
+          val keyBytes = message.getGetInitialState.getValue.toByteArray
+          // The key row is serialized as a byte array, we need to convert it back to a Row
+          val keyRow = PythonSQLUtils.toJVMRow(keyBytes, groupingKeySchema, keyRowDeserializer)
+          val groupingKeyToInternalRow =
+            ExpressionEncoder(groupingKeySchema).createSerializer().apply(keyRow)
+          val iter = initialStateKeyToRowMap
+            .get(groupingKeyToInternalRow).getOrElse(Iterator.empty)
+
+          var seenInitStateOnKey = false
+          while (iter.hasNext) {
```

Comment:
IIUC, we only process the first item in the iterator. If that's the case, why do we need to use arrow to send that single row?

Also, it seems the initial state map is `Map[InternalRow, Iterator[InternalRow]]`. Why do we have multiple value rows mapping to a single key row? Is that because we allow a pandas dataframe as an input of initial state on python side?

Reply:
> Also, it seems the initial state map is Map[InternalRow, Iterator[InternalRow]]. Why do we have multiple value rows mapping to a single key row? Is that because we allow a pandas dataframe as an input of initial state on python side?

Yes, we allow the input of initial state to have duplicate rows of the same grouping key as users input their spark dataframe as initial state in the stateful processor here; to avoid memory overflow, the spark dataframe are computed lazily; so we won't know which row is duplicated on the grouping key until we process till the duplicated row in the iterator is processed.